### ISSUE TITLE
Comparision with Official implementation

### ISSUE THREAD

**Issue by amil-rp-work** [lshug/first-order-model-tf#10]
@lshug Wow wonderful job on porting from Pytorch to TF. I can see a lot of work done by you here!!!
I had 2 questions

1.  Are the results the same as the official ones?
2. Is there any speed up in inferencing time? I used the example.ipynb file however in the second iteration of animate it takes a Very Long time. Any ideas?

**Comment by lshug** [comment]
Hello.


1. 
Mean absolute difference between the float32 representations of the output images of the original and of my version when animating with the same source image and driving video is less than 10^-5. In practice this translates to the outputs being identical. I treat significant differences between the original's output and the output of my model on the same data and with the same modes as bugs, so if you find any differences in outputs between the implementations, please open up an issue.


2. 

I'm assuming by first iteration (first tqdm progress bar) you mean the kp detection stage, and by the second iteration (second tqdm progress bar) you mean the prediction generation stage.

I address this in the FAQ. Most of the inference time in both the kp detector and the generator spent on conv2d's, so it's unlikely that code optimization will improve the performance significantly. On my laptop's crappy old i7 it takes about 3.8-4.2 seconds per four-frame batch. If you need significantly better performance, you'll have to either run on a machine with a GPU or make a heavily-optimized custom tflite conversion.

One thing you try is increase the batch size, which will slightly improve per-frame performance but will also increase memory requirement. Another thing you can do is build the savedmodels with --predictiononly arg, which will build the generator in such a way that it only outputs the predicted frame, without the other stuff in its structured output that's used in visualization mode and in training (mask, sparse_deformed, occlusion_map, and the deformed source image). Neither of those options will halve or come anywhere close to halving the per-batch inference time though.

For significant optimization, you'll need to do custom tflite conversion. I put links to the relevant pages on TF guides in the FAQ section I mentioned above.

**Comment by amil-rp-work** [comment]
So @lshug  I am currently running on GPU and I get 7its/s which is too slow. For stats, the same input/output on PyTorch runs for about 10 secs and on TF it runs for 2 mins?
I am currently using batch_size = 1. But even increasing it results in the same time.

**Comment by lshug** [comment]
Are you sure your numbers are correct? If you're processing 7 frames per second for a total of two minutes on TF, and are doing the same on PyTorch in 10 seconds, then pytorch version must be processing 84 frames per second, which seems above the capability of the model. It'd mean that PyTorch version is outputting 35 seconds worth of frames in 10 seconds, which is thee times the real-time processing speed, and I haven't seen first order model run that fast on a GPU. Can you replicate on Google Colab?

**Comment by lshug** [comment]
I made a [Colab for measuring the latency of both versions on GPU](https://colab.research.google.com/drive/1CHlfG792RifIpwYQqqUNHGS_3rUe16vQ?usp=sharing). The two modes seem to run with around the same latency.
